File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/LinkParsingSample.java
package org.pdfclown.samples.cli;

import java.awt.geom.Rectangle2D;
import java.util.List;
import java.util.Map;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.PageAnnotations;
import org.pdfclown.documents.contents.ITextString;
import org.pdfclown.documents.files.FileSpecification;
import org.pdfclown.documents.interaction.actions.Action;
import org.pdfclown.documents.interaction.actions.GoToDestination;
import org.pdfclown.documents.interaction.actions.GoToEmbedded;
import org.pdfclown.documents.interaction.actions.GoToEmbedded.PathElement;
import org.pdfclown.documents.interaction.actions.GoToNonLocal;
import org.pdfclown.documents.interaction.actions.GoToURI;
import org.pdfclown.documents.interaction.annotations.Annotation;
import org.pdfclown.documents.interaction.annotations.Link;
import org.pdfclown.documents.interaction.navigation.document.Destination;
import org.pdfclown.files.File;
import org.pdfclown.objects.PdfObjectWrapper;
import org.pdfclown.tools.TextExtractor;
import org.pdfclown.util.io.IOUtils;

/**
  This sample demonstrates <b>how to inspect the links of a PDF document</b>, retrieving
  their associated text along with its graphic attributes (font, font size, text color,
  text rendering mode, text bounding box...).
  <h3>Remarks</h3>
  <p>According to PDF spec, page text and links have no mutual relation (contrary to, for
  example, HTML links), so retrieving the text associated to a link is somewhat tricky
  as we have to infer the overlapping areas between links and their corresponding text.</p>

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.8
  @version 0.1.2.1, 04/08/15
*/
public class LinkParsingSample
  extends Sample
{
  @Override
  public void run(
    )
  {
    File file = null;
    try
    {
      // 1. Opening the PDF file...
      {
        String filePath = promptFileChoice("Please select a PDF file");
        try
        {file = new File(filePath);}
        catch(Exception e)
        {throw new RuntimeException(filePath + " file access error.",e);}
      }
      Document document = file.getDocument();

      // 2. Link extraction from the document pages.
      TextExtractor extractor = new TextExtractor();
      extractor.setAreaTolerance(2); // 2 pt tolerance on area boundary detection.
      boolean linkFound = false;
      for(Page page : document.getPages())
      {
        if(!promptNextPage(page, !linkFound))
        {
          quit();
          break;
        }

        Map<Rectangle2D,List<ITextString>> textStrings = null;
        linkFound = false;

        // Get the page annotations!
        PageAnnotations annotations = page.getAnnotations();
        if(!annotations.exists())
        {
          System.out.println("No annotations here.");
          continue;
        }

        // Iterating through the page annotations looking for links...
        for(Annotation<?> annotation : annotations)
        {
          if(annotation instanceof Link)
          {
            linkFound = true;

            if(textStrings == null)
            {textStrings = extractor.extract(page);}

            Link link = (Link)annotation;
            Rectangle2D linkBox = link.getBox();

            // Text.
            /*
              Extracting text superimposed by the link...
              NOTE: As links have no strong relation to page text but a weak location correspondence,
              we have to filter extracted text by link area.
            */
            StringBuilder linkTextBuilder = new StringBuilder();
            for(ITextString linkTextString : extractor.filter(textStrings,linkBox))
            {linkTextBuilder.append(linkTextString.getText());}
            System.out.println("Link '" + linkTextBuilder + "' ");

            // Position.
            System.out.println(
              "    Position: "
                + "x:" + Math.round(linkBox.getX()) + ","
                + "y:" + Math.round(linkBox.getY()) + ","
                + "w:" + Math.round(linkBox.getWidth()) + ","
                + "h:" + Math.round(linkBox.getHeight())
                );

            // Target.
            System.out.print("    Target: ");
            PdfObjectWrapper<?> target = link.getTarget();
            if(target instanceof Destination)
            {printDestination((Destination)target);}
            else if(target instanceof Action)
            {printAction((Action)target);}
            else if(target == null)
            {System.out.println("[not available]");}
            else
            {System.out.println("[unknown type: " + target.getClass().getSimpleName() + "]");}
          }
        }
        if(!linkFound)
        {
          System.out.println("No links here.");
          continue;
        }
      }
    }
    finally
    {
      // 3. Closing the PDF file...
      IOUtils.closeQuietly(file);
    }
  }

  private void printAction(
    Action action
    )
  {
    System.out.println("Action [" + action.getClass().getSimpleName() + "] " + action.getBaseObject());
    if(action instanceof GoToDestination<?>)
    {
      if(action instanceof GoToNonLocal<?>)
      {
        FileSpecification<?> destinationFile = ((GoToNonLocal<?>)action).getDestinationFile();
        if(destinationFile != null)
        {System.out.println("      Filename: " + destinationFile.getPath());}

        if(action instanceof GoToEmbedded)
        {
          PathElement target = ((GoToEmbedded)action).getDestinationPath();
          System.out.println("      EmbeddedFilename: " + target.getEmbeddedFileName() + " Relation: " + target.getRelation());
        }
      }
      System.out.print("      ");
      printDestination(((GoToDestination<?>)action).getDestination());
    }
    else if(action instanceof GoToURI)
    {System.out.println("      URI: " + ((GoToURI)action).getURI());}
  }

  private void printDestination(
    Destination destination
    )
  {
    System.out.println(destination.getClass().getSimpleName() + " " + destination.getBaseObject());
    System.out.print("        Page ");
    Object pageRef = destination.getPage();
    if(pageRef instanceof Page)
    {
      Page page = (Page)pageRef;
      System.out.println(page.getNumber() + " [ID: " + page.getBaseObject() + "]");
    }
    else
    {System.out.println(((Integer)pageRef+1));}

    Object location = destination.getLocation();
    if(location != null)
    {System.out.println(String.format("        Location %s", location));}

    Double zoom = destination.getZoom();
    if(zoom != null)
    {System.out.println(String.format("        Zoom %s", zoom));}
  }
}

File: PdfClown-NetStandard/CliSamples/org/pdfclown/samples/cli/LinkParsingSample.cs
using org.pdfclown.bytes;
using org.pdfclown.documents;
using org.pdfclown.documents.files;
using org.pdfclown.documents.contents;
using actions = org.pdfclown.documents.interaction.actions;
using org.pdfclown.documents.interaction.annotations;
using org.pdfclown.documents.interaction.navigation.document;
using files = org.pdfclown.files;
using org.pdfclown.objects;
using org.pdfclown.tools;

using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to inspect the links of a PDF document, retrieving
    their associated text along with its graphic attributes (font, font size, text color,
    text rendering mode, text bounding box...).</summary>
    <remarks>According to PDF spec, page text and links have no mutual relation (contrary to, for
    example, HTML links), so retrieving the text associated to a link is somewhat tricky
    as we have to infer the overlapping areas between links and their corresponding text.</remarks>
  */
  public class LinkParsingSample
    : Sample
  {
    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(var file = new files::File(filePath))
      {
        Document document = file.Document;
  
        // 2. Link extraction from the document pages.
        TextExtractor extractor = new TextExtractor();
        extractor.AreaTolerance = 2; // 2 pt tolerance on area boundary detection.
        bool linkFound = false;
        foreach(Page page in document.Pages)
        {
          if(!PromptNextPage(page, !linkFound))
          {
            Quit();
            break;
          }

          IDictionary<RectangleF?,IList<ITextString>> textStrings = null;
          linkFound = false;
  
          // Get the page annotations!
          PageAnnotations annotations = page.Annotations;
          if(!annotations.Exists())
          {
            Console.WriteLine("No annotations here.");
            continue;
          }
  
          // Iterating through the page annotations looking for links...
          foreach(Annotation annotation in annotations)
          {
            if(annotation is Link)
            {
              linkFound = true;
  
              if(textStrings == null)
              {textStrings = extractor.Extract(page);}
  
              Link link = (Link)annotation;
              RectangleF linkBox = link.Box;
  
              // Text.
              /*
                Extracting text superimposed by the link...
                NOTE: As links have no strong relation to page text but a weak location correspondence,
                we have to filter extracted text by link area.
              */
              StringBuilder linkTextBuilder = new StringBuilder();
              foreach(ITextString linkTextString in extractor.Filter(textStrings,linkBox))
              {linkTextBuilder.Append(linkTextString.Text);}
              Console.WriteLine("Link '" + linkTextBuilder + "' ");
  
              // Position.
              Console.WriteLine(
                "    Position: "
                  + "x:" + Math.Round(linkBox.X) + ","
                  + "y:" + Math.Round(linkBox.Y) + ","
                  + "w:" + Math.Round(linkBox.Width) + ","
                  + "h:" + Math.Round(linkBox.Height)
                  );
  
              // Target.
              Console.Write("    Target: ");
              PdfObjectWrapper target = link.Target;
              if(target is Destination)
              {PrintDestination((Destination)target);}
              else if(target is actions::Action)
              {PrintAction((actions::Action)target);}
              else if(target == null)
              {Console.WriteLine("[not available]");}
              else
              {Console.WriteLine("[unknown type: " + target.GetType().Name + "]");}
            }
          }
          if(!linkFound)
          {
            Console.WriteLine("No links here.");
            continue;
          }
        }
      }
    }

    private void PrintAction(
      actions::Action action
      )
    {
      /*
        NOTE: Here we have to deal with reflection as a workaround
        to the lack of type covariance support in C# (so bad -- any better solution?).
      */
      Console.WriteLine("Action [" + action.GetType().Name + "] " + action.BaseObject);
      if(action.Is(typeof(actions::GoToDestination<>)))
      {
        if(action.Is(typeof(actions::GotoNonLocal<>)))
        {
          FileSpecification destinationFile = (FileSpecification)action.Get("DestinationFile");
          if(destinationFile != null)
          {Console.WriteLine("      Filename: " + destinationFile.Path);}

          if(action is actions::GoToEmbedded)
          {
            actions::GoToEmbedded.PathElement target = ((actions::GoToEmbedded)action).DestinationPath;
            Console.WriteLine("      EmbeddedFilename: " + target.EmbeddedFileName + " Relation: " + target.Relation);
          }
        }
        Console.Write("      ");
        PrintDestination((Destination)action.Get("Destination"));
      }
      else if(action is actions::GoToURI)
      {Console.WriteLine("      URI: " + ((actions::GoToURI)action).URI);}
    }

    private void PrintDestination(
      Destination destination
      )
    {
      Console.WriteLine(destination.GetType().Name + " " + destination.BaseObject);
      Console.Write("        Page ");
      object pageRef = destination.Page;
      if(pageRef is Page)
      {
        Page page = (Page)pageRef;
        Console.WriteLine(page.Number + " [ID: " + page.BaseObject + "]");
      }
      else
      {Console.WriteLine((int)pageRef+1);}

      object location = destination.Location;
      if(location != null)
      {Console.WriteLine("        Location {0}", location);}

      double? zoom = destination.Zoom;
      if(zoom.HasValue)
      {Console.WriteLine("        Zoom {0}", zoom.Value);}
    }
  }
}